libs/Microsoft.MixedReality.WebRTC.Native/src/peer_connection.cpp

```diff
@@ -1274,63 +1670,620 @@ void PeerConnectionImpl::OnLocalDescCreated(
+  } else {
+    RTC_DCHECK(IsPlanB());
+    // In Plan B, since there is no guarantee about the order of tracks, they
+    // are matched by stream ID (msid).
+    int mline_index = -1;
+    std::string name;
+    if (!ExtractMlineIndexAndNameFromReceiverPlanB(receiver, mline_index,
+                                                   name)) {
+      return Error(
+          Result::kUnknownError,
+          "Failed to associate RTP receiver with Plan B emulated mline index "
+          "for track pairing.");
+    }
+
+    // Create a new audio transceiver wrapper for it
+    return CreateTransceiverPlanB(media_kind, mline_index, std::move(name),
+                                  receiver);
+  }
+}
+
+ErrorOr<RefPtr<Transceiver>> PeerConnectionImpl::GetOrCreateTransceiver(
+    int mline_index,
+    webrtc::RtpTransceiverInterface* rtp_transceiver,
+    std::string name) {
+  // Convert and validate media kind
+  mrsMediaKind media_kind;
+  switch (rtp_transceiver->media_type()) {
+    case cricket::MediaType::MEDIA_TYPE_AUDIO:
+      media_kind = mrsMediaKind::kAudio;
+      break;
+    case cricket::MediaType::MEDIA_TYPE_VIDEO:
+      media_kind = mrsMediaKind::kVideo;
+      break;
+    default:
+      return Error(Result::kInvalidMediaKind, "Unknown SDP semantic");
+  }
+
+  // Find an existing transceiver wrapper which would have been created just
+  // a moment ago by the remote track added callback.
+  rtc::CritScope lock(&tracks_mutex_);
+  for (auto&& tr : transceivers_) {
+    if (tr && (tr->impl() == rtp_transceiver)) {
+      RTC_DCHECK(media_kind == tr->GetMediaKind());
+      return static_cast<RefPtr<Transceiver>>(tr);
+    }
+  }
+  // Not found - create a new one
+  return CreateTransceiverUnifiedPlan(media_kind, mline_index, std::move(name),
+                                      rtp_transceiver);
+}
+
+ErrorOr<RefPtr<Transceiver>> PeerConnectionImpl::GetOrCreateTransceiverPlanB(
+    int mline_index,
+    webrtc::RtpReceiverInterface* rtp_receiver,
+    std::string name) {
+  // Convert and validate media kind
+  mrsMediaKind media_kind;
+  switch (rtp_receiver->media_type()) {
+    case cricket::MediaType::MEDIA_TYPE_AUDIO:
+      media_kind = mrsMediaKind::kAudio;
+      break;
+    case cricket::MediaType::MEDIA_TYPE_VIDEO:
+      media_kind = mrsMediaKind::kVideo;
+      break;
+    default:
+      return Error(Result::kInvalidMediaKind, "Unknown SDP semantic");
+  }
+
+  // Find an existing transceiver wrapper which would have been created just
+  // a moment ago by the remote track added callback.
+  rtc::CritScope lock(&tracks_mutex_);
+  for (auto&& tr : transceivers_) {
+    if (tr && tr->HasReceiver(rtp_receiver)) {
+      RTC_DCHECK(media_kind == tr->GetMediaKind());
+      return static_cast<RefPtr<Transceiver>>(tr);
+    }
+  }
+  // Not found - create a new one
+  return CreateTransceiverPlanB(media_kind, mline_index, std::move(name),
+                                rtp_receiver);
+}
```

Comment:
nit: these might be better as one single FindTransceiver method that returns null if not found + a call to CreateTransceiver(UnifiedPlan|PlanB) at the point of call.

Reply:
Fixed. Also moved some code inline when the method is only called once.